File: lib/PDLx/bin_on_index.pl
#!perl
## no critic ProhibitAccessOfPrivateData

use Types::Common::Numeric qw[ PositiveNum PositiveInt PositiveOrZeroInt ];
use Types::Standard qw[ Optional InstanceOf slurpy Dict Bool Enum Undef Int ];
use Type::Params qw[ compile ];

use Carp;
use PDL::Lite;

my $bin_on_index_check;


BEGIN {

    $bin_on_index_check = compile(
        slurpy Dict [
            index      => InstanceOf ['PDL'],
            signal     => Optional   [ InstanceOf ['PDL'] ],
            nbins      => InstanceOf ['PDL'] | PositiveInt,
            error      => Optional   [ InstanceOf ['PDL'] | Undef ],
            error_algo => Optional   [ Enum [ keys %MapErrorAlgo ] | Undef ],
            oob        => Optional   [ Bool ],
            offset     => Optional   [Int],
        ] );
}

sub bin_on_index {

    my ( $opts ) = $bin_on_index_check->( @_ );

    # specify defaults
    my %opt = (
       oob => 0,
        %$opts,
    );

    $opt{error_algo} ||=
        defined $opt{signal} && defined $opt{error} ? 'sdev'
      : defined $opt{signal}                        ? 'sdev'
      : defined $opt{error}                         ? 'rss'
      :                                               'poisson';

    if ( $opt{oob} ) {

	$opt{offset} = 1 unless defined $opt{offset};
	$opt{nbins} +=2;
    }

    $opt{offset} = 0 unless defined $opt{offset};

    croak( "must specify error attribute if 'rss' errors selected\n" )
      if !defined $opt{error} && $opt{error_algo} eq 'rss';

    $opt{flags}
      = ( ( defined $opt{signal}      && BIN_ARG_HAVE_SIGNAL ) || 0 )
      | ( ( defined $opt{error}       && BIN_ARG_HAVE_ERROR  ) || 0 )
      | ( ( $opt{set_bad}             && BIN_ARG_SET_BAD     ) || 0 )
      | ( ( $opt{oob}                 && BIN_ARG_SAVE_OOB    ) || 0 )
      | $MapErrorAlgo{ $opt{error_algo} };

    $opt{maxnbins} = PDL::Core::topdl( $opt{nbins} )->max;

    my @pin   = qw[ signal index error nbins ];
    my @pout  = qw[ nelem b_signal b_error b_mean ];
    my @ptmp  = qw[ b_error2 b_m2 b_weight  ];
    my @oargs = qw[ flags maxnbins offset ];


    # several of the input piddles are optional.  the PP routine
    # doesn't know that and will complain about the wrong number of
    # dimensions if we pass a null piddle. A 1D zero element piddle
    # will have its dimensions auto-expanded without much
    # wasted memory.
    $opt{signal} = PDL->zeroes( PDL::double, 0 ) if !defined $opt{signal};

    # make sure we use the same type as the input signal
    $opt{$_} = $opt{signal}->zeroes( 1 ) for grep { !defined $opt{$_} } @pin;

    $opt{$_} = PDL->null for grep { !defined $opt{$_} } @pout;
    $opt{$_} = PDL->null for grep { !defined $opt{$_} } @ptmp;

    _bin_on_index_int( @opt{ @pin, @pout, @ptmp, @oargs } );

    my %results = map {
        ( my $nkey = $_ ) =~ s/^b_//;
        $nkey, $opt{$_}
    } @pout;


    return \%results;
}


=pod

=head2 bin_on_index

=for usage

  $hashref = bin_on_index( %pars  );

=for ref

Bin data and optional errors using an existing piddle which provides
the bin index for each datum.

This routine ignores data with bad values or with errors that have
bad values.

=head3 Parameters

B<bin_on_index> is passed a hash or a reference to a hash containing
its parameters.  The possible parameters are:

=over

=item C<index> I<piddle>

The bin index. Valid bins indices range from C<0> to C<nbins - 1>.
The index may be offset using the C<offset> parameter.  The
C<oob_algo> parameter specifies how out-of-bounds indices are handled.
Required.

=item C<nbins>  I<integer|piddle>

The number of bins, including any used for out-of-bound data. Required.

=item C<signal> I<piddle>

A piddle containing the signal data.  If not
specified, each datum is given a value of C<1>. Optional.


=item C<error> I<piddle>

A piddle with the error for signal datum. Optional.

=item C<error_algo> I<string>

How the error is to be handled or calculated.  It
may have one of the following values:

=over

=item * C<poisson>

Poisson errors will be caculated based upon the number of elements in a bin,

  error**2 = N

Any input errors are ignored.

=item * C<sdev>

The error is the population standard deviation of the signal in a bin.

  error**2 = Sum [ ( signal - mean ) **2 ] / ( N - 1 )

If errors are provided, they are used to calculated the weighted population
standard deviation.

  error**2 = ( Sum [ (signal/error)**2 ] / Sum [ 1/error**2 ] - mean**2 )
             * N / ( N - 1 )

=item * C<rss>

Errors must be provided; the errors of elements in a bin are added in
quadrature.

=back

The default value depends upon which data are available:

  signal   error   default
  ------   -----   -------
     Y       Y       sdev
     Y       N       sdev
     N       Y       rss
     N       N       poisson


=item C<offset> I<integer>

An offset to be added to the index.  This is useful for transforming
input index values so that they lie within C<[0,nbins-1]>  For example,
if indices for in-bounds data range from I<0> to I<N>, and the input
indices use I<-1> and I<N+1> to indicate data are out-of-bounds,
then setting C<nbins> to I<N+2> and C<offset> to C<0> will accumulate
the out-of-bounds data into bins with indices C<0> and C<N+1>

This defaults to C<0>.

=item C<oob> I<boolean>

If an input index (after addition of C<offset> ) is less than I<0>
or greater than I<nbins-1>, it is out-of-bounds and cannot be used.  If
C<oob> is false, (the default) those data are ignored.

If C<oob> is true then indices less than C<0> are set to C<0>
and indices greater than C<nbins-1> are set to C<nbins-1>.  If C<oob> is
true the following parameter adjustments are made:

=over

=item *

C<nbins> is increased by two to accomodate the extra bins

=item *

If C<offset> has not been set, it is set to C<1>.

=back




=back


=head3 Results

B<bin_on_index> returns a hashref with the following entries:

=over

=item C<signal>

A piddle containing the sum of the signal in each bin.

=item C<nelem>

A piddle containing the number of data elements in each bin.

=item C<error>

A piddle containing the errors in each bin, calculated using the
algorithm specified via C<error_algo>.

=item C<mean>

A piddle containing the possibly weighted mean of the signal in each
bin.

=back

=cut

File: lib/PDLx/bin_on_index.c
<% if ( $PDL_BAD_CODE ) {'
#undef PDL_BAD_CODE
#define PDL_BAD_CODE
'}%>


int flags = $COMP(optflags);

int have_signal   = flags & BIN_ARG_HAVE_SIGNAL;
int have_error    = flags & BIN_ARG_HAVE_ERROR;
int error_sdev    = flags & BIN_ARG_ERROR_SDEV;
int error_poisson = flags & BIN_ARG_ERROR_POISSON;
int error_rss     = flags & BIN_ARG_ERROR_RSS;
int save_oob      = flags & BIN_ARG_SAVE_OOB;

threadloop %{

    <% $PDL_Indx %> nbins_m1 = $nbins() - 1;
    <% $PDL_Indx %> offset = $COMP(offset);
  <%
    # intialize output and temp bin data one at a time to
    # avoid trashing the cache
    join( "\n", map {
	         "loop(nb) %{ \$${_}() = 0; %}"
	        }   qw/ nelem b_signal b_mean
                        b_error2 b_m2 b_weight /
    );
  %>

  /* if we could preset min & max to the initial value in a bin,
     we could shave off a comparison.  Unfortunately, we can't
     do that, as we can't know apriori which is the first
     element in a bin. */
  loop(nb) %{ $b_min() =  DBL_MAX; %}
  loop(nb) %{ $b_max() = -DBL_MAX; %}


  loop(n) %{

    <% $PDL_Indx %> nelem;
    <% $PDL_Indx %> index = $index();

    double signal = have_signal ? $signal(nsig => n ) : 1;
    double error;
    double error2;
    double weight;
    double bweight;

#ifdef PDL_BAD_CODE
    if (   $ISBADVAR(signal,signal)
        || $ISBADVAR(index,index)
         ) {
      continue;
    }
#endif /* PDL_BAD_CODE */

    index += offset;


    if ( save_oob ) {

	if ( index < 0 )             index = 0;
	else if ( index > nbins_m1 ) index = nbins_m1;

    }
    else if ( index < 0 || index > nbins_m1 )
	continue;

    nelem = ++$nelem( nb => index );
    $b_signal(nb => index)  += signal;

    if ( signal < $b_min( nb => index ) ) $b_min( nb => index ) = signal;
    if ( signal > $b_max( nb => index ) ) $b_max( nb => index ) = signal;

    if ( have_error ) {

      error = $error();

#ifdef PDL_BAD_CODE
    if ( $ISBADVAR(error,error) ) {
      continue;
    }
#endif /* PDL_BAD_CODE */

      error2 = error * error;
      weight = 1 / error2;
      bweight = $b_weight( nb => index ) += weight;

    }
    else {

	weight = 1;
	bweight = nelem;
    }


    {
	/** stable means of generating the mean */
	double delta = signal - $b_mean( nb => index );
	double bmean = $b_mean( nb => index)  += delta * weight / bweight;

	/* calculate error */
	if ( error_sdev ) {

	    /* incremental algorithm for possibly weighted standard deviation; see
	       https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
	    */

	    $b_m2( nb => index ) += weight * delta * ( signal - bmean );

	}

	else if ( error_rss ) {

	    $b_error2( nb => index )   += error2;

	}

    }

  %}


  /* summarize */
 loop (nb) %{

     if ( error_poisson ) {


       $b_error() = sqrt( $nelem() );
       $b_mean() = $b_signal() / $nelem();
     }

     else if ( error_sdev ) {

	 <% $PDL_Indx %> nelem = $nelem();
	 int bad_error = nelem <= 1;

	 double norm = have_error ?  nelem / $b_weight() : 1;

	 $b_error( ) = bad_error
	     ? DBL_MAX
	     : sqrt( $b_m2() * norm / (nelem - 1) );

     }

     else {

 	  $b_error() = sqrt( $b_error2() );
     }

     %}
%}

<% if ( $PDL_BAD_CODE ) {'
#undef PDL_BAD_CODE
'} %>
